core-to-plc/src/Language/Plutus/CoreToPLC/Definitions.hs

```diff
@@ -0,0 +1,119 @@
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Language.Plutus.CoreToPLC.Definitions where
+
+import           Language.Plutus.CoreToPLC.Error
+import           Language.Plutus.CoreToPLC.PLCTypes
+
+import qualified Language.PlutusCore                as PLC
+import           Language.PlutusCore.MkPlc
+
+import qualified GhcPlugins                         as GHC
+
+import           Control.Monad.Except
+
+import           Data.Foldable
+import qualified Data.Graph                         as Graph
+import qualified Data.Map                           as Map
+
+-- | The visibility of a definition. See Note [Abstract data types]
+data Visibility = Abstract | Visible
+-- | A definition of type 'val' with variable type 'var'.
+data Def var val = Def {dVis::Visibility, dVar::var, dVal::val}
+
+isVisible :: Def var val -> Bool
+isVisible (Def vis _ _) = case vis of
+    Abstract -> False
+    Visible  -> True
+
+-- | Either a simple type or a datatype with constructors and a matcher.
+data TypeRep = PlainType PLCType | DataType PLCType [TermDef] TermDef
+
+trTy :: TypeRep -> PLCType
+trTy = \case
+    PlainType t -> t
+    DataType t _ _ -> t
+
+type TypeDef = Def PLCTyVar TypeRep
+
+tydTy :: TypeDef -> PLCType
+tydTy = \case
+    Def Abstract (n, _) _ -> PLC.TyVar () n
+    Def Visible _ tr -> trTy tr
+
+tydConstrs :: TypeDef -> Maybe [TermDef]
+tydConstrs = \case
+    Def _ _ (DataType _ constrs _) -> Just constrs
+    _ -> Nothing
+
+tydMatch :: TypeDef -> Maybe TermDef
+tydMatch = \case
+    Def _ _ (DataType _ _ match) -> Just match
+    _ -> Nothing
+
+type TermDef = Def PLCVar PLCTerm
+
+tdTerm :: TermDef -> PLCTerm
+tdTerm = \case
+    Def Abstract (n, _) _ -> PLC.Var () n
+    Def Visible _ t -> t
+
+type DefMap key def = Map.Map key (def, [key])
+
+-- Processing definitions
+
+data TermOrTypeDef = TermDef TermDef | TypeDef TypeDef
+
+-- | Given the definitions in the program, create a topologically ordered list of the SCCs using the dependency information
+defSccs :: DefMap GHC.Name TypeDef -> DefMap GHC.Name TermDef -> [Graph.SCC TermOrTypeDef]
+defSccs typeDefs termDefs =
+    let
+        typeInputs = fmap (\(ghcName, (d, deps)) -> (TypeDef d, ghcName, deps)) (Map.assocs typeDefs)
+        termInputs = fmap (\(ghcName, (d, deps)) -> (TermDef d, ghcName, deps)) (Map.assocs termDefs)
+    in
+        -- weirdly this produces them in reverse topological order
+        reverse $ Graph.stronglyConnComp (typeInputs ++ termInputs)
+
+wrapWithDefs
+    :: (MonadError ConvError m)
+    => DefMap GHC.Name TypeDef
+    -> DefMap GHC.Name TermDef
+    -> PLC.Term PLC.TyName PLC.Name ()
+    -> m (PLC.Term PLC.TyName PLC.Name ())
+wrapWithDefs typeDefs termDefs body = do
+    let sccs = defSccs typeDefs termDefs
+    -- process from the inside out
+    foldM wrapDefScc body (reverse sccs)
```

Comment:
`reverse` something that has just been `reverse`d?

Reply:
Yes, this is weird, but I think it's clearer to produce them in the right order and then reverse them where we "really" reverse them. I'm hoping I'll be able to switch to a more sensible graph library, at which point there will be only one reverse.